src/utilities/include/vec_dh.h

```diff
@@ -40,244 +125,195 @@ namespace manifold {
+      if (this->size_ > 0)
+        uninitialized_copy(autoPolicy(this->size_), this->ptr_,
+                           this->ptr_ + this->size_, newBuffer);
+      if (this->ptr_ != nullptr) {
         TracyFreeS(ptr_, 3);
-        free(ptr_);
+        free(this->ptr_);
       }
-      ptr_ = newBuffer;
+      this->ptr_ = newBuffer;
       capacity_ = n;
     }
   }
 
   void resize(int newSize, T val = T()) {
-    bool shrink = size_ > 2 * newSize;
+    bool shrink = this->size_ > 2 * newSize;
     reserve(newSize);
-    if (size_ < newSize) {
-      uninitialized_fill(autoPolicy(newSize - size_), ptr_ + size_,
-                         ptr_ + newSize, val);
+    if (this->size_ < newSize) {
+      uninitialized_fill(autoPolicy(newSize - this->size_),
+                         this->ptr_ + this->size_, this->ptr_ + newSize, val);
     }
-    size_ = newSize;
+    this->size_ = newSize;
     if (shrink) shrink_to_fit();
   }
 
   void shrink_to_fit() {
     T *newBuffer = nullptr;
-    if (size_ > 0) {
-      newBuffer = reinterpret_cast<T *>(malloc(size_ * sizeof(T)));
+    if (this->size_ > 0) {
+      newBuffer = reinterpret_cast<T *>(malloc(this->size_ * sizeof(T)));
       if (newBuffer == nullptr) throw std::bad_alloc();
       TracyAllocS(newBuffer, size_ * sizeof(T), 3);
-      uninitialized_copy(autoPolicy(size_), ptr_, ptr_ + size_, newBuffer);
-      if (ptr_ != nullptr) {
+      uninitialized_copy(autoPolicy(this->size_), this->ptr_,
+                         this->ptr_ + this->size_, newBuffer);
+      if (this->ptr_ != nullptr) {
         TracyFreeS(ptr_, 3);
-        free(ptr_);
+        free(this->ptr_);
       }
     }
-    ptr_ = newBuffer;
-    capacity_ = size_;
+    this->ptr_ = newBuffer;
+    capacity_ = this->size_;
   }
 
-#ifdef MANIFOLD_DEBUG
-  void Dump() const {
-    std::cout << "VecDH = " << std::endl;
-    for (int i = 0; i < size_; ++i) {
-      std::cout << i << ", " << ptr_[i] << ", " << std::endl;
+  VecDHView<T> get_view(int offset = 0, int length = -1) {
+    if (length == -1) {
+      length = this->size_ - offset;
+      if (length < 0) throw std::out_of_range("VecDH::get_view out of range");
+    } else if (offset + length > this->size_ || offset < 0) {
+      throw std::out_of_range("VecDH::get_view out of range");
+    } else if (length < 0) {
+      throw std::out_of_range("VecDH::get_view negative length is not allowed");
     }
-    std::cout << std::endl;
+    return VecDHView<T>(this->ptr_ + offset, length);
   }
-#endif
 
- private:
-  int size_ = 0;
-  int capacity_ = 0;
-  T *ptr_ = nullptr;
-};
-
-template <typename T>
-class VecDc {
- public:
-  VecDc(const VecDH<T> &vec) : ptr_(vec.ptrD()), size_(vec.size()) {}
-
-  __host__ __device__ const T &operator[](int i) const { return ptr_[i]; }
-  __host__ __device__ int size() const { return size_; }
-
- private:
-  T const *const ptr_;
-  const int size_;
-};
+  VecDHView<const T> get_cview(int offset = 0, int length = -1) const {
```

Comment:
Since we use these a lot, let's go for a short name. How about just `view/cview`?

Reply:
sounds good. actually we can overload implicit conversion for the majority of cases, I wonder if you like that or not

Reply:
Overloads are fine, I just like the const to be clear for readability, but that is already specified in the functor, which is clearer anyway.